os_dep.c

```diff
@@ -2251,7 +2239,7 @@ GC_register_data_segments(void)
   /* Avoid even referencing DATASTART and DATAEND as they are   */
   /* unnecessary and cause linker errors when bitcode is        */
   /* enabled.  GC_register_data_segments is not called anyway.  */
-#  elif defined(DYNAMIC_LOADING) && defined(DARWIN)
+#  elif defined(DYNAMIC_LOADING) && (defined(DARWIN) || defined(HAIKU))
   /* No-op.  GC_register_main_static_data() always returns false. */
```

Comment:
GC_register_main_static_data() return constant false only for Darwin and some FreeBSD variants (unless USE_PROC_FOR_LIBRARIES). But not for Haiku.

Reply:
On Haiku, the main application binary is also a "shared image" and will be reported in an `image_info` same as any dynamically loaded libraries.

Reply:
OK, I see. Then we should define `GC_register_main_static_data() { return FALSE; }` (like in Darwin case).